### PR TITLE
[compiler-v2] Bytecode verification at the end of compilation

### PR DESCRIPTION
**Pull request by vineethk**
### Description

Closes https://github.com/aptos-labs/aptos-core/issues/11868 (along with the promised bonus points).

In this PR, we add bytecode verification after generating file-format bytecode in compiler v2 (as well as in the tests). We also add the file-format bytecode output to the debug log (eg., when compiler v2 is run with `MVC_LOG=debug`).

I have filed two related issues: https://github.com/aptos-labs/aptos-core/issues/11924 (exposed due to adding bytecode verification sanity check, more on this below) and https://github.com/aptos-labs/aptos-core/issues/11925 (provide better location information on bytecode verification failure).

### Test Plan

All tests run as before, except four test cases in file-format-generator. These test failures (due to bytecode verification failures) were because of missing ability annotations, which I have now added to those tests, "correcting" the tests. I have also preserved those missing ability tests in their original form, but in the ability-checker directory (where it is supposed to be), exposing bugs in the ability checker.

I have manually verified that `MVC_LOG=debug` prints the bytecode as expected.

I have also added a new test folder and (incorrect) pipeline to trigger the bytecode verification failure.

### CODE REVIEW

File: third_party/move/move-compiler-v2/src/lib.rs [hunk 1]
```diff
@@ -232,6 +245,30 @@ fn add_default_optimization_pipeline(pipeline: &mut FunctionTargetPipeline) {
...
+        .iter()
+        .map(|unit| {
+            let view = match unit {
+                CompiledUnit::Module(module) => BinaryIndexedView::Module(&module.module),
+                CompiledUnit::Script(script) => BinaryIndexedView::Script(&script.script),
+            };
+            Disassembler::from_view(view, location::Loc::new(FileHash::empty(), 0, 0))
+                .and_then(|d| d.disassemble())
+        })
+        .collect();
```
Comment (brmataptos) [inline third_party/move/move-compiler-v2/src/lib.rs:260]: Why not do the `?` here?  The var type will be simplified.

**Review by brmataptos — COMMENTED** [review]
I'm not seeing whether we have tests cases showing a bytecode validation failure, to make sure the plumbing is correct.  Also, if we can suppress stackless bytecode dumps on cases illustrating validation failures the tests will be less brittle.

**Comment by vineethk** [comment]
> I'm not seeing whether we have tests cases showing a bytecode validation failure, to make sure the plumbing is correct.

I did think about this, and it is a bit weird: a test case showing a bytecode validation failure is a compiler bug, so eventually, these should all be fixed. So maintaining such a test is weird IMO. The finding of four failures in existing test cases is a demo of this plumbing, in a way.

> Also, if we can suppress stackless bytecode dumps on cases illustrating validation failures the tests will be less brittle.

Yes, I did think about this as I was looking through the new errors for the ported test cases. I would consider it out of scope for this PR, but can follow it up - it needs to be done more widely.

**Comment by rahxephon89** [comment]
> I'm not seeing whether we have tests cases showing a bytecode validation failure, to make sure the plumbing is correct. Also, if we can suppress stackless bytecode dumps on cases illustrating validation failures the tests will be less brittle.

Do we have code examples that can pass compiler but fail in bytecode verifier (the one caused by a compiler bug does not count)?

File: third_party/move/move-compiler-v2/src/lib.rs (hunk 1 above)
Comment (vineethk) [reply third_party/move/move-compiler-v2/src/lib.rs:260]: Not sure I understand what you are saying here? Use `?` where?

I am using the pattern from here: https://doc.rust-lang.org/rust-by-example/error/iter_result.html#fail-the-entire-operation-with-collect

**Comment by brmataptos** [comment]
> > I'm not seeing whether we have tests cases showing a bytecode validation failure, to make sure the plumbing is correct.
> 
> I did think about this, and it is a bit weird: a test case showing a bytecode validation failure is a compiler bug, so eventually, these should all be fixed. So maintaining such a test is weird IMO. The finding of four failures in existing test cases is a demo of this plumbing, in a way.

We could have a test flag and/or compiler pass to inject something invalid.

File: third_party/move/move-compiler-v2/src/lib.rs (hunk 1 above)
Comment (brmataptos) [reply third_party/move/move-compiler-v2/src/lib.rs:260]: Yes, but you have a `?` on the following use of this value to return the error immediately. Why not here?

**Comment by vineethk** [comment]
> Do we have code examples that can pass compiler but fail in bytecode verifier (the one caused by a compiler bug does not count)?

All code examples that can pass a compiler but fail in bytecode verifier, by definition, should be a compiler bug (in fact, we use compiler `Bug` as the severity to report bytecode verification failure).

**Comment by vineethk** [comment]
> We could have a test flag and/or compiler pass to inject something invalid.

Yes, we can certainly "test" this "internal compiler error" by injecting something invalid, making it conditional on a flag. But I would argue this is no different than other internal compiler errors that we have, and we do not inject errors to check them - I am not convinced of the benefits of the added complexity.

File: third_party/move/move-compiler-v2/src/lib.rs (hunk 1 above)
Comment (vineethk) [reply third_party/move/move-compiler-v2/src/lib.rs:260]: In order to use the pattern I linked above. Maybe you can show me a compilable version of what you are suggesting?

**Comment by vineethk** [comment]
@brmataptos I have tried to address your comments, PTAL.

I ended up adding a new test folder and corresponding (incorrect) pipeline to showcase the bytecode verification failure.

**Review by brmataptos — APPROVED** [review]
Nice.  I thinking of suggesting that we interpret and pointing to the error location, but since it most likely represents a compiler bug that's proabbly not as useful as telling us about it if someone sees this.

File: third_party/move/move-compiler-v2/src/lib.rs (hunk 1 above)
Comment (vineethk) [reply third_party/move/move-compiler-v2/src/lib.rs:260]: The change you suggest above is not compilable.

Comment (brmataptos) [reply third_party/move/move-compiler-v2/src/lib.rs:260]: hmm.  Well, never mind, then.

**Comment by vineethk** [comment]
> I thinking of suggesting that we interpret and pointing to the error location, but since it most likely represents a compiler bug that's proabbly not as useful as telling us about it if someone sees this.

As referenced in the PR description, I have a "low" priority bug opened for this: https://github.com/aptos-labs/aptos-core/issues/11925

**Review by fEst1ck — APPROVED** [review]
LGTM

**Pull request merged**